IPython/core/completer.py

```diff
@@ -2360,11 +2360,122 @@ def _jedi_matches(
             else:
                 return iter([])
 
+    class _CompletionContextType(enum.Enum):
+        ATTRIBUTE = "attribute"  # For attribute completion
+        GLOBAL = "global"  # For global completion
+
+    def _determine_completion_context(self, line):
+        """
+        Determine whether the cursor is in an attribute or global completion context.
+        """
+        # Cursor in string/comment → GLOBAL.
+        if self._is_in_string_or_comment(line):
+            return self._CompletionContextType.GLOBAL
+
+        # Match 'anything-dot-word' at end for attribute context.
+        # Ex: 'obj.', 'np.random.ran'.
+        chain_match = re.search(r"([a-zA-Z_].*)\.([a-zA-Z_][a-zA-Z0-9_]*)?$", line)
+        if chain_match:
+            return self._CompletionContextType.ATTRIBUTE
+
+        # No attribute pattern → GLOBAL.
+        # Ex: 'var', '', '3.', '-42.5.'
+        return self._CompletionContextType.GLOBAL
+
+    def _is_in_string_or_comment(self, text):
```

Comment:
This will fail on nested f-strings:

![image](https://github.com/user-attachments/assets/3a120c31-c79b-4824-9321-59792cd60108)

I am not sure if we want to support it (if it is easy then sure, if it is hard - not sure), but at least need to document it, possibly by adding xfail test?